src/main/kotlin/org/neo4j/graphql/BuildingEnv.kt

```diff
@@ -0,0 +1,197 @@
+package org.neo4j.graphql
+
+import graphql.schema.*
+
+class BuildingEnv(val types: MutableMap<String, GraphQLType>) {
+
+    private val typesForRelation = types.values
+        .filterIsInstance<GraphQLObjectType>()
+        .filter { it.getDirective(DirectiveConstants.RELATION) != null }
+        .map { it.getDirectiveArgument<String>(DirectiveConstants.RELATION, DirectiveConstants.RELATION_NAME, null)!! to it.name }
+        .toMap()
+
+    fun buildFieldDefinition(
+            prefix: String,
+            resultType: GraphQLOutputType,
+            scalarFields: List<GraphQLFieldDefinition>,
+            nullableResult: Boolean,
+            forceOptionalProvider: (field: GraphQLFieldDefinition) -> Boolean = { false }
+    ): GraphQLFieldDefinition.Builder {
+        var type: GraphQLOutputType = resultType
+        if (!nullableResult) {
+            type = GraphQLNonNull(type)
+        }
+        return GraphQLFieldDefinition.newFieldDefinition()
+            .name("$prefix${resultType.name}")
+            .arguments(getInputValueDefinitions(scalarFields, forceOptionalProvider))
+            .type(type.ref() as GraphQLOutputType)
+    }
+
+    fun getInputValueDefinitions(
+            relevantFields: List<GraphQLFieldDefinition>,
+            forceOptionalProvider: (field: GraphQLFieldDefinition) -> Boolean): List<GraphQLArgument> {
+        return relevantFields.map { field ->
+            var type = field.type as GraphQLType
+            type = getInputType(type)
+            type = if (forceOptionalProvider.invoke(field)) {
+                (type as? GraphQLNonNull)?.wrappedType ?: type
```

Comment:
shouldn't there be an extension function for this?

Reply:
since it is operation on a list, I don't see the right cutting point here to extract the function as extension function.